cpp/src/join/conditional_join_kernels.cuh

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <join/join_common_utils.cuh>
+#include <join/join_common_utils.hpp>
+
+#include <cub/cub.cuh>
+#include <cudf/ast/detail/linearizer.hpp>
+#include <cudf/ast/detail/transform.cuh>
+#include <cudf/ast/operators.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/table/table_device_view.cuh>
+
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Computes the output size of joining the left table to the right table.
+ *
+ * This method uses a nested loop to iterate over the left and right tables and count the number of
+ * matches according to a boolean expression.
+ *
+ * @tparam block_size The number of threads per block for this kernel
+ * @tparam has_nulls Whether or not the inputs may contain nulls.
+ *
+ * @param[in] left_table The left table
+ * @param[in] right_table The right table
+ * @param[in] JoinKind The type of join to be performed
+ * @param[in] compare_nulls Controls whether null join-key values should match or not.
+ * @param[in] plan Container of device data required to evaluate the desired expression.
+ * @param[out] output_size The resulting output size
+ */
+template <int block_size, bool has_nulls>
+__global__ void compute_conditional_join_output_size(table_device_view left_table,
+                                                     table_device_view right_table,
+                                                     join_kind JoinKind,
+                                                     null_equality compare_nulls,
+                                                     ast::detail::device_ast_plan plan,
+                                                     cudf::size_type* output_size)
+{
+  // The (required) extern storage of the shared memory array leads to
+  // conflicting declarations between different templates. The easiest
+  // workaround is to declare an arbitrary (here char) array type then cast it
+  // after the fact to the appropriate type.
+  extern __shared__ char raw_intermediate_storage[];
```

Comment:
This approach, FWIW, is idiomatic in CUDA C++.

Reply:
I've frequently encountered this approach for handling multiple arrays of different types in the same kernel, but I've never personally written or seen templated kernels where the template parameter was intended to control the type of the shared memory array. The issue was easy enough to diagnose when I saw the compiler error, but I didn't realize the problem until then, so I thought the comment would be helpful for future readers. Do you think it's superfluous and can be removed?